browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+  addFencedAction(action);
+};
+
+const flood = () => {
+  const args = {
+    label: getModel().foreground,
+    x_location: getCanvas().imgX,
+    y_location: getCanvas().imgY,
+  };
+  const action = new BackendAction('flood', args);
+  addFencedAction(action);
+};
+
+const trim = () => {
+  const args = {
+    label: getCanvas().label,
+    frame: getModel().frame,
+    x_location: getCanvas().imgX,
+    y_location: getCanvas().imgY,
+  };
+  const action = new BackendAction('trim_pixels', args);
+  addFencedAction(action);
+};
+
+const erode = () => {
+  const args = {
+    label: getCanvas().label
+  };
+  const action = new BackendAction('erode', args);
+  addFencedAction(action);
+};
+
+const dilate = () => {
+  const args = {
+    label: getCanvas().label
+  };
+  const action = new BackendAction('dilate', args);
+  addFencedAction(action);
+};
+
+const autofit = () => {
+  const args = {
+    label: getCanvas().label
+  };
+  const action = new BackendAction('active_contour', args);
+  addFencedAction(action);
+};
+
+const predictFrame = () => {
+  const args = {
+    frame: getModel().frame
+  };
+  const action = new BackendAction('predict_single', args);
+  addFencedAction(action);
+};
+
+const predictAll = () => {
+  const args = {};
+  const action = new BackendAction('predict_zstack', args);
+  addFencedAction(action);
+};
+
+const swapFrame = () => {
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background,
+    frame: getModel().frame
+  };
+  const action = new BackendAction('swap_single_frame', args);
+  addFencedAction(action);
+};
+
+const replaceFrame = () => {
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background,
+    // frame: getModel().frame? may need to add if we no longer store frame on backend
+  };
+  const action = new BackendAction('replace_single', args);
+  addFencedAction(action);
+};
+
+const replaceAll = () => {
+  const args = {
+    label_1: getModel().foreground,
+    label_2: getModel().background
+  };
+  const action = new BackendAction('replace', args);
+  addFencedAction(action);
+};
+
+const watershed = (context) => {
+  const args = {
+    frame: getModel().frame,
+    label: context.storedLabel,
+    x1_location: context.storedX,
+    y1_location: context.storedY,
+    x2_location: getCanvas().imgX,
+    y2_location: getCanvas().imgY
+  };
+  const action = new BackendAction('watershed', args);
+  addFencedAction(action);
+};
+
+// tool states
+const paintState = {
+  initial: 'idle',
+  entry: assign({ trace: [] }),
+  states: {
+    idle: {
+      on: {
+        mousedown: {
+          // prevents select and paint at the same time
+          cond: (_, event) => !event.shiftKey,
+          target: 'dragging',
+          actions: 'addToTrace'
+        }
+      }
+    },
+    dragging: {
+      on: {
+        mousemove: { actions: 'addToTrace' },
+        mouseup: { actions: 'draw' },
```

Comment:
In `thresholdState` we use a second object here: `{ target: 'idle' },` which I assume puts the state back into `idle` after the first action is finished.

Why do we use the `target` there but not here?

Reply:
Hmmm, I think we don't need the transition for either state, as both the `draw` action and the `threshold` action will result in an `EDIT`, `LOADING`, and `LOADED` events that will cause the mouse state to transition in & out of the `loaded` state and then to the toolbar History state to return to `paint.idle` or `threshold.idle`. However, I think it would be better to add in the `{ target: 'idle' }` because we go to the initial or idle state when we use a shallow history, but not a deep history. By always transitioning to `idle`, both the shallow and deep history would behave the same.

Reply:
Just took a second look at the second object `{ target: 'idle' }` on threshold. This is actually a fallback transition when for when we lift the mouse and the threshold bounding box is empty, so we can't threshold. When the box is nonempty, we do the `threshold` action (and don't transition to idle), and otherwise transition back to `idle` so we can draw another bounding box. Based on my previous comment, I think we should add transitions to `idle` for both the paint state and threshold state when we launch the `paint` or `threshold` actions.